### DISCUSSION TITLE
Path object boolean

### DISCUSSION THREAD

**Discussion by vrafaeli** [friction2d/friction#200] [Q&A] [answered]
I wanted to make a simple hollow object.
The process as I know with many other graphics tools:
Make two primary objects, stack one on top another, then run them to one of the vector boolean operations.

For the result object to be hollow, the "mask" part is to be placed completely inside/within the bigger/exterior part.

However using Friction this doesn't result in a hollow object.
You get a hole if the Paths option in the Properties is set to Even-odd instead of Winding. I have no idea what this setting is about; never saw them in other tools.

This setting works in the viewport and video render. When exported to svg, the hole isn't there.

Grouping the two primary objects into a layer with one of them used as mask (DestOut) is worse. Both objects just show up as is/are when exported to svg.

Out of curiosity I repeat the same routine using Inkscape (path boolean, not masking), which exports just fine.
When opened in a web browser, the inspection tool (F12) shows they seem to have different approach in the svg notation. I won't get too much into that since the format is too complicated for me.

On closer look:
The Winding mode may result in mask cutout effect, but only if at least one of the nodes of the smaller object strays outside the bigger object's geometry, if that makes any sense. 

Then the two individual components, previously the primary objects, are to be set to different modes in the Properties dialog box, between Remove Reverse, Exclusion, Divide, a combo that I cannot memorize. If all the small object's nodes are within the big object, the hollow bit just disappears.

But bottom line is the vector graphics in Friction can still be improved. I tested with Enve too—same result. Looks like it's always been that way and has never been addressed?

**Comment by rodlie** [comment]
Thanks for reporting, I usually prepare vectors in Inkscape before animating them in Friction, so I might overlook issues from time to time.

If you want to make a donut (circle in circle) then 'even-odd' is needed for the resulting paths, but the svg exporter does not add `fill-rule` for 'even-odd' (https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/fill-rule), this is a bug. Other than that, I'm unsure, as 'winding' and 'even-odd' does what they are supposed to do(?) (https://en.wikipedia.org/wiki/Even%E2%80%93odd_rule) as far as I know (please correct me if I'm wrong).

**Reply by rodlie** [reply, marked as answer]
Snapshot build with support for `even-odd` on SVG export: https://sourceforge.net/projects/friction/files/0.9.6.1-47660298/

**Reply by vrafaeli** [reply]
Thanks for looking into the issue.
Downloaded the new build, made a quick test and exported it back to Friction.
So the `fill-rule="evenodd"` works charms. This is great! I read up a bit more on the svg stuff.

Codewise the result is still quite jumbled with the grouping overkill. I guess one can always do a manual cleanup (I just did on the test object), but to most artists it might not be worth the time to shave off a mere few kilobytes.

Another thing perhaps more important than cleanup is that the fill color isn't retained when the svg is imported back in.
It works when it uses template color for the fill such as green or blue, but using rgba format it just shows up all black. Previously it doesn't have a color (alpha is 1 but no fill), I add something like `version="1"` to the `<svg>`. Not sure if that's helpful. Should this be a separate thread? Would be great if something can be done about this; real-life designers working with their client's sophisticated logo would need it.
